model/task_lifecycle.go

```diff
@@ -806,202 +796,243 @@ func evalStepback(t *task.Task, caller, status string, deactivatePrevious bool)
-	if err != nil {
-		return errors.WithStack(err)
+	githubBuildStatus := getBuildStatus(githubStatusTasks)
+
+	if githubBuildStatus == b.GithubCheckStatus {
+		return nil
 	}
-	if b == nil {
-		return errors.Errorf("build '%s' not found", b.Id)
+
+	if evergreen.IsFinishedBuildStatus(githubBuildStatus) {
+		event.LogBuildGithubCheckFinishedEvent(b.Id, githubBuildStatus)
 	}
 
-	buildTasks, err := task.Find(task.ByBuildId(b.Id))
+	return b.UpdateGithubCheckStatus(githubBuildStatus)
+}
+
+// UpdateBuildStatus updates the status of the build based on its tasks' statuses.
+func UpdateBuildStatus(b *build.Build) error {
+	buildTasks, err := task.Find(task.ByBuildId(b.Id).WithFields(task.StatusKey, task.ActivatedKey, task.DependsOnKey, task.IsGithubCheckKey))
 	if err != nil {
-		return errors.WithStack(err)
+		return errors.Wrapf(err, "getting tasks in build '%s'", b.Id)
 	}
 
-	githubCheckTasks := map[string]bool{}
-	githubCheckStatus := ""
-	successfulGithubCheckTasks := 0
-	// also check the current status of github checks
-	for _, bt := range buildTasks {
-		if bt.IsGithubCheck {
-			githubCheckTasks[bt.Id] = bt.IsGithubCheck
-		}
+	buildStatus := getBuildStatus(buildTasks)
+
+	if buildStatus == b.Status {
+		return nil
 	}
 
-	failedTask := false
-	finishedTasks := 0
-	blockedTasks := 0
-	unfinishedTasks := 0
+	event.LogBuildStateChangeEvent(b.Id, buildStatus)
 
-	cache := task.NewDisplayTaskCache()
-	// update the build's status based on tasks for this build
-	for _, t := range buildTasks {
-		if !t.IsFinished() {
-			if t.Blocked() {
-				blockedTasks++
-			} else if utility.StringSliceContains(evergreen.UncompletedStatuses, t.Status) {
-				unfinishedTasks++
-			}
-			continue
+	if evergreen.IsFinishedBuildStatus(buildStatus) {
+		if err = b.MarkFinished(buildStatus, time.Now()); err != nil {
+			return errors.Wrapf(err, "marking build as finished with status '%s'", buildStatus)
 		}
-		finishedTasks++
-
-		var displayTask *task.Task
-		displayTask, err = cache.Get(&t)
-		if err != nil {
-			return errors.WithStack(err)
+		if err = updateMakespans(b, buildTasks); err != nil {
+			return errors.Wrapf(err, "updating makespan information for '%s'", b.Id)
 		}
-		if displayTask != nil {
-			t = *displayTask
+	} else {
+		if err = b.UpdateStatus(buildStatus); err != nil {
+			return errors.Wrap(err, "updating build status")
 		}
+	}
 
-		// update the build's status when a test task isn't successful
-		if evergreen.IsFailedTaskStatus(t.Status) {
-			if err = b.UpdateStatus(evergreen.BuildFailed); err != nil {
-				return errors.Wrap(err, "Error updating build status")
-			}
-			if githubCheckTasks[t.Id] {
-				githubCheckStatus = evergreen.BuildFailed
-			}
-			failedTask = true
-		} else if githubCheckTasks[t.Id] {
-			successfulGithubCheckTasks++
+	if err = updateBuildGithubStatus(b, buildTasks); err != nil {
+		return errors.Wrap(err, "updating build github status")
+	}
+
+	return nil
+}
+
+func getVersionStatus(builds []build.Build) string {
```

Comment:
Hmm. My read of John's comment is that he preferred to avoid the situation that triggered the bug in EVG-14117 instead of directly addressing it because the fix he had in mind would have been a performance hit.
I don't think this PR will fix EVG-14117: the build with only blocked tasks will still never finish.